src/tabpanel.c

```diff
@@ -0,0 +1,616 @@
+/* vi:set ts=8 sts=4 sw=4 noet:
+ *
+ * VIM - Vi IMproved	by Bram Moolenaar
+ *
+ * Do ":help uganda"  in Vim to read copying and usage conditions.
+ * Do ":help credits" in Vim to see a list of people who contributed.
+ * See README.txt for an overview of the Vim source code.
+ */
+
+/*
+ * tabpanel.c:
+ */
+
+#include "vim.h"
+
+#if defined(FEAT_TABPANEL) || defined(PROTO)
+
+static void do_by_tplmode(int tplmode, int col_start, int col_end,
+	int* pcurtab_row, int* ptabpagenr);
+
+// set pcurtab_row. don't redraw tabpanel.
+#define TPLMODE_GET_CURTAB_ROW	0
+// set ptabpagenr. don't redraw tabpanel.
+#define TPLMODE_GET_TABPAGENR	1
+// redraw tabpanel.
+#define TPLMODE_REDRAW		2
+
+#define TPL_FILLCHAR		' '
+
+// tpl_wrap's values
+#define WRAP_OFF		0
+#define WRAP_ON			1
+// tpl_align's values
+#define ALIGN_LEFT		0
+#define ALIGN_RIGHT		1
+
+static int tpl_wrap = WRAP_OFF;
+static int tpl_align = ALIGN_LEFT;
+static int tpl_columns = 20;
+static char tpl_vert = '\0';
+
+
+typedef struct {
+    win_T   *wp;
+    win_T   *cwp;
+    char_u  *user_defined;
+    int	maxrow;
+    int	offsetrow;
+    int	*prow;
+    int	*pcol;
+    int	attr;
+    int	col_start;
+    int	col_end;
+} tabpanel_T;
+
+    int
+tabpanelopt_changed(void)
+{
+    char_u	*p;
+
+    tpl_wrap = WRAP_OFF;
+    tpl_align = ALIGN_LEFT;
+    tpl_columns = 20;
+    tpl_vert = '\0';
+
+    p = p_tplo;
+    while (*p != NUL)
+    {
+	if (STRNCMP(p, "wrap", 4) == 0)
+	{
+	    p += 4;
+	    tpl_wrap = WRAP_ON;
+	}
+	else if (STRNCMP(p, "align:left", 10) == 0)
+	{
+	    p += 10;
+	    tpl_align = ALIGN_LEFT;
+	}
+	else if (STRNCMP(p, "align:right", 11) == 0)
+	{
+	    p += 11;
+	    tpl_align = ALIGN_RIGHT;
+	}
+	else if (STRNCMP(p, "columns:", 8) == 0 && VIM_ISDIGIT(p[8]))
+	{
+	    p += 8;
+	    tpl_columns = getdigits(&p);
+	}
+	else if (STRNCMP(p, "vert:", 5) == 0)
+	{
+	    p += 5;
+	    if (*p != NUL && *p != ',')
+	    {
+		tpl_vert = *p;
+		++p;
+	    }
+	}
+
+	if (*p != ',' && *p != NUL)
+	    return FAIL;
```

Comment:
when this returns fail, the partly some options are already set. Better to use some temporary options and only once we know there is no FAILURE in parsing the option string, set the option values to the new values (from those temporary options).